### PR TITLE
Add overlay until option

### PR DESCRIPTION
**Pull request by sgallag-insta**
**Summary of changes:**

- Added `--overlay_until` option.
  - Takes an integer argument that is the number of iterations.
  - Default value is `None`.

Tests can be run by executing `python -m unittest tests/test_pixray.py` from the main pixray directory.

### CODE REVIEW

**Comment by dribnet** [comment]
I understand the logic of `sys.maxsize` and python3 is already a dependency, but I'm wondering for consistency with the other arguments what your thoughts are on having the default be `None` instead and then checking for that explicitly?

**Comment by sgallag-insta** [comment]
Either approach is fine in my mind. I just went with the former since it was a little more concise but there is certainly something to be said for consistency. I can update it!

**Comment by sgallag-insta** [comment]
Updated. I also took the opportunity to clean up my tests a bit. Previously it was handling arguments set to `None` rather lazily.

**Comment by dribnet** [comment]
Also would be more consistent if `overlay-until` is provided as a percentage instead of a raw iteration count to be consistent with other args (eg: `--learning_rate_drops` and `vdiff_skip`). However I can think of use cases for both, so perhaps we could consider coming up with a standard that we eventually propagate to other args.

proposal 1:
--
Just make overlay_until a percentage and merge it in (don't overthink it!)

proposal 2:
--
also allow non-integers related to iteration count which will get resolved to an integer somewhere in `process_args`. examples:
  "20%"
  "5iters"
This "iteration count specifier" parsing could live in utils.py like the color parsing currently does for plug ins (like vdiff).

Note that this rationale for how to handle this could also be repurposed elsewhere - in fact color parsing has already been standardised internally in this way and is used already by multiple modules. For example, edge loss currently uses a percentage of the width/height when configuring the border size, but could also allow "5px" etc.

**Comment by sgallag-insta** [comment]
Gotcha! Actually, it seems there are some further inconsistencies here. My thought was to implement it as an iteration number to make it consistent with `overlay_every`. I guess that should probably have a percentage option too.

I think the string parsing approach sounds pretty reasonable. Personally, I would advocate that argument handling might warrant it's own class/file but I suppose that doesn't need to happen _right_ now. I will update it :+1:

**Comment by dribnet** [comment]
Yep - agreed it's a bit of a confusing mess. Just hoping not to make it worse. 😅

**Comment by dribnet** [comment]
Yep - agreed it's a bit of a confusing mess. Just hoping not to make it worse. 😅 

But given that `overlay_every` (and `overlay_offset`) is already in raw iterations, I now agree you are right and it makes more sense to merge this as-is since those are the most related to `overlay_until`. So how about we go forward with this version and perhaps in the future we can standardize raw numbers to be iteration counts and percentages to be explicitly provided via "20%".

**Comment by sgallag-insta** [comment]
Sounds good to me!

**Comment by dribnet** [comment]
Just tested this and it's working great. Merging into `future`.

One of the original use cases I had for `overlay` was do do a sort of feature visualization which adds details to an original image. So starting from an image of a bee like this:

bee

In the old days one would just use this as an `init_image`, but doing that by itself is way too powerful and quickly diverges from the original.

```
python pixray.py \
  --quality normal \
  --aspect square \
  --label 'bee' \
  --init_noise snow \
  --init_image_alpha 200 \
  --init_image inputs/bee.jpg \
  --save_every 20 \
  --output outputs/jan22/beeref_29_03.png
```
beeref_29_03

So the `overlay` was meant to keep it closer to the original in a crude kind of way. Here's what that looks like:

```
python pixray.py \
  --quality normal \
  --aspect square \
  --label 'bee' \
  --init_noise snow \
  --init_image_alpha 200 \
  --init_image inputs/bee.jpg \
  --overlay_image inputs/bee.jpg \
  --save_every 20 \
  --overlay_every 10 \
  --overlay_alpha 128 \
  --learning_rate_drops 200 \
  --output outputs/jan22/beeref_29_04.png
```
beeref_29_04

Kind of crude, but you get the idea. Note that there is a weird `--learning_rate_drops 200` thrown in which effectively disables the learning rate. That is because if you don't do this, then the overlay will overpower the process at the end after the learning rate has been turned down. So:

```
python pixray.py \
  --quality normal \
  --aspect square \
  --label 'bee' \
  --init_noise snow \
  --init_image_alpha 200 \
  --init_image inputs/bee.jpg \
  --overlay_image inputs/bee.jpg \
  --save_every 20 \
  --overlay_every 10 \
  --overlay_alpha 128 \
  --output outputs/jan22/beeref_29_01.png
```
beeref_29_01

But now with `overlay_until` one can just stop the overlay before the learning rate drops, which leads to more imagined details being added after that point.

```
python pixray.py \
  --quality normal \
  --aspect square \
  --label 'bee' \
  --init_noise snow \
  --init_image_alpha 200 \
  --init_image inputs/bee.jpg \
  --overlay_image inputs/bee.jpg \
  --save_every 20 \
  --overlay_every 10 \
  --overlay_alpha 128 \
  --overlay_until 160 \
  --output outputs/jan22/beeref_29_02.png
```

beeref_29_02

Probably not the best example in the world - but just documenting a bit one of the ways this can be used.

**Comment by sgallag-insta** [comment]
Thanks for the great write-up! This should probably go in the docs somewhere. I had some idea that `overlay_until` might be useful for avoiding the noise caused by starting a new generation but I didn't realize it's relationship with the learning rate. I definitely have to spend some more time playing around with this!

**Comment by dribnet** [comment]
Yeah - not the best place for it but wanted to brain dump it somewhere as I was testing.

Thanks also for your addition of the test code. If you are at all interested in helping me administer the repo going forward let me know as I would welcome help from anyone that is knowledgable about adding tests, reviewing pull requests, and planning releases.
